src/adapter.jl

```diff
@@ -15,8 +15,22 @@ function Base.string( obj::Union{GapObj,FFE} )
 end
 
 ## implement indexing interface
-Base.getindex(x::GapObj, i::Int64) = Globals.ELM_LIST(x, i)
-Base.setindex!(x::GapObj, v::Any, i::Int64 ) = Globals.ASS_LIST( x, i, v )
+Base.getindex(x::GapObj, i::Integer) = Globals.ELM_LIST(x, Int(i))
+Base.setindex!(x::GapObj, v, i::Integer) = Globals.ASS_LIST( x, Int(i), v)
+
+function Base.iterate(x::GapObj)
+    Globals.IsList(x) || throw(ArgumentError("GAP object is not iterable:\n $x"))
+    len = length(x)
+    iszero(len) && return nothing
+    return x[1], (1, len)
+end
+function Base.iterate(x::GapObj, state)
+    s, len = state
+    s == len && return nothing
+    return x[s+1], (s+1, len)
+end
```

Comment:
What about the following extension:
```
function Base.iterate(x::GapObj)
    if Globals.IsList(x)
      len = length(x)
      iszero(len) && return nothing
      return x[1], (1, len)
    elseif Globals.IsDomain(x)
      iter = Globals.Iterator( x )
      Globals.IsDoneIterator( iter ) && return nothing
      return Globals.NextIterator( iter ), iter
    else
      throw(ArgumentError("GAP object is not iterable:\n $x"))
    end
end
function Base.iterate(x::GapObj, state::Tuple{Integer,Integer})
    s, len = state
    s == len && return nothing
    return x[s+1], (s+1, len)
end
function Base.iterate(x::GapObj, state::GapObj)
    Globals.IsDoneIterator( state ) && return nothing
    return Globals.NextIterator( state ), state
end
```
Then one can do the following.
```
g = GAP.Globals.SymmetricGroup( 3 )
for i in GAP.Globals.Elements( g ) print( i, ", " ) end
for i in g print( i, ", " ) end
```

Reply:
this was just need to write concise code for constructors;

it does work, but indeed this iterate(Gapobj) is a julia anti-pattern, even if we can infer eltype of list and group. There is little point in writing more type unstable code, even if it works. If we know that a `GapObj` represents a group, lets wrap it in a `GAP(Perm)group` type and iterate over it in a type-stable form.

@fingolfin I just had an idea (since I believe your long-term plan is to reduce the amount of magic conversions in GAP) to explicitly set a `no-change-to-representation-please`-bitflag on objects coming from julia. This is a way to 'catch-them-all' (I mean all functions that try to do so) and eventually replace/rewrite them (they will fail on such objects). But maybe my understanding is too simplistic anyway ;)

Reply:
I like what @ThomasBreuer wrote.

@kalmarek could you elaborate in how far this is a Julia anti-pattern?

And yeah, that `no-change-to-representation-please` object flag is precisely what I had in mind during our Skype call (I'd call it "locked representation object flag", which reflects existing GAP terminology). But I don't think it's very important right now, we have more pressing issues :-)

Reply:
its an anti-pattern since if You write `for g in G ...`, `g` will be of type `Union{Nothing, Int, BigInt, GapObj, Julia wrapped thing, Mptr...}` which essentially means `Any`.
If You create `PList{T, ElT}(::GapObj{T}) where {T, ElT}` first, this introduces a "function barrier" (where type instability is contained) and then iterate on `PList` in a type stable way. Whenever you have a switch statement to decide what to return (that cannot be decided at compile time), you're venturing into the world of `Any`.

Reply:
in the sense I see no point in adding more code like this, until a design decision (on how to deal with Gap types on the julia side) is arrived at